### ISSUE TITLE
Pana assumes that the git root is the package directory

### ISSUE THREAD

**Issue by tjarvstrand** [dart-lang/pana#1224]
Currently, Pana seems to assume that if there is a git root above the package directory, then that should be used as the base for the analysis. It then proceeds to copy this entire directory to a temporary location.

Unfortunately, this is very unsuitable for use in monorepo settings. Our repository contains 10's of thousands of files, totalling many gigabytes, making pana essentially unusable by default.

It's possible to work around that limitation by manually copying the package to a temporary location outside the git tree but at the moment there is no information that the copy operation is happening so I had to dig through the source to figure out what was going on.

There should be a way to override this behavior or, at the very least, something informing the user about the ongoing copying.

**Comment by isoos** [comment]
Yes, that's correct, pana does copy the content by default, because we may change a few things while running the analysis, and it is not desired to have side-effects on the original code, only on the working copy.

Can you give us more info about the project layout? E.g. how the packages depend on each other (assuming `path:` dependencies), or can the large files excluded with some pattern?

**Comment by tjarvstrand** [comment]
> Yes, that's correct, pana does copy the content by default, because we may change a few things while running the analysis, and it is not desired to have side-effects on the original code, only on the working copy.

This is totally fine and understandable, but pana shouldn't automatically assume that the entire repository should be copied.

> Can you give us more info about the project layout? E.g. how the packages depend on each other (assuming path: dependencies), or can the large files excluded with some pattern?

In terms of the package itself, the structure is a fairly simple Flutter plugin, with a native component. It does however live in the same Git repository as several other (non-flutter/Dart) packages and intermediate steps in the build process pull in and build large external dependencies (e.g. webRTC).

**Comment by isoos** [comment]
I could imagine a flag that disables the git-root detection, effectively making a copy only on the package directory. However, when the package has any `path: ../other_package` dependency, the analysis will fail in such cases. wdyt @jonasfj?

**Comment by sigurdm** [comment]
We could try to go towards not copying the package at all.

One way would be to create a stub-package in a temp-folder with a single dependency on the analyzed package and do resolution there.

We would probably have to run `pub outdated --no-dev-dependencies` in the analyzed project folder.

**Comment by sigurdm** [comment]
My proposed solution would perhaps make it hard to override the analysis_options (because they are always found in the enclosing file system).
